### ISSUE TITLE
No eligible torsions found during geometry engine proposal

### ISSUE THREAD

**Issue by jchodera** [choderalab/perses#114]
See #109:
```

======================================================================
ERROR: Testing MultiTargetDesign sampler with alanine dipeptide mutation transfer free energy from vacuum -> implicit
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/choderaj/miniconda/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/Users/choderaj/github/choderalab/perses/perses.choderalab/perses/samplers/samplers.py", line 934, in run
    self.update()
  File "/Users/choderaj/github/choderalab/perses/perses.choderalab/perses/samplers/samplers.py", line 919, in update
    self.update_samplers()
  File "/Users/choderaj/github/choderalab/perses/perses.choderalab/perses/samplers/samplers.py", line 889, in update_samplers
    sampler.update()
  File "/Users/choderaj/github/choderalab/perses/perses.choderalab/perses/samplers/samplers.py", line 811, in update
    self.update_sampler()
  File "/Users/choderaj/github/choderalab/perses/perses.choderalab/perses/samplers/samplers.py", line 785, in update_sampler
    self.sampler.update()
  File "/Users/choderaj/github/choderalab/perses/perses.choderalab/perses/samplers/samplers.py", line 673, in update
    self.update_sampler()
  File "/Users/choderaj/github/choderalab/perses/perses.choderalab/perses/samplers/samplers.py", line 635, in update_sampler
    geometry_new_positions, geometry_logp  = self.geometry_engine.propose(topology_proposal, ncmc_old_positions, self.sampler.thermodynamic_state.beta)
  File "/Users/choderaj/github/choderalab/perses/perses.choderalab/perses/rjmc/geometry.py", line 112, in propose
    torsion, logp_choice = self._choose_torsion(atoms_with_positions, atom)
  File "/Users/choderaj/github/choderalab/perses/perses.choderalab/perses/rjmc/geometry.py", line 546, in _choose_torsion
    torsion_idx = np.random.randint(0, len(eligible_torsions))
  File "mtrand.pyx", line 945, in mtrand.RandomState.randint (numpy/random/mtrand/mtrand.c:10732)
ValueError: low >= high
```
Looks like this might be triggered when the new molecule does not have any eligible torsions.  Will try to track down exactly what kinds of transformations cause this to happen.

**Comment by jchodera** [comment]
From #109, it seems like this failure occurs only for ring-containing atoms:
```
======================================================================
ERROR: Testing MultiTargetDesign sampler with alkanes mutation transfer free energy from vacuum -> implicit
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/choderaj/miniconda/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/Users/choderaj/github/choderalab/perses/perses.choderalab/perses/samplers/samplers.py", line 940, in run
    self.update()
  File "/Users/choderaj/github/choderalab/perses/perses.choderalab/perses/samplers/samplers.py", line 925, in update
    self.update_samplers()
  File "/Users/choderaj/github/choderalab/perses/perses.choderalab/perses/samplers/samplers.py", line 895, in update_samplers
    sampler.update()
  File "/Users/choderaj/github/choderalab/perses/perses.choderalab/perses/samplers/samplers.py", line 817, in update
    self.update_sampler()
  File "/Users/choderaj/github/choderalab/perses/perses.choderalab/perses/samplers/samplers.py", line 791, in update_sampler
    self.sampler.update()
  File "/Users/choderaj/github/choderalab/perses/perses.choderalab/perses/samplers/samplers.py", line 679, in update
    self.update_sampler()
  File "/Users/choderaj/github/choderalab/perses/perses.choderalab/perses/samplers/samplers.py", line 641, in update_sampler
    geometry_new_positions, geometry_logp  = self.geometry_engine.propose(topology_proposal, ncmc_old_positions, self.sampler.thermodynamic_state.beta)
  File "/Users/choderaj/github/choderalab/perses/perses.choderalab/perses/rjmc/geometry.py", line 112, in propose
    torsion, logp_choice = self._choose_torsion(atoms_with_positions, atom)
  File "/Users/choderaj/github/choderalab/perses/perses.choderalab/perses/rjmc/geometry.py", line 547, in _choose_torsion
    raise Exception("No eligible torsions found for placing atom %s." % str(atom_for_proposal))
Exception: No eligible torsions found for placing atom <Atom CD1 [16]; In PHE 1>.
```
It could be that only impropers torsions are defined for this atom.

We should figure out if this is the case---that only impropers are defined for some ring atoms---and if so, how we can either
* (1) use impropers instead
* (2) create our own list of torsions (`i-j-k-l` topology-bonded atoms), since we are no longer tied to using explicit `PeriodicTorsionForce` entries in `System`; this would allow us to extend to CHARMM as well

**Comment by pgrinaway** [comment]
Ok, I think using topological torsions will probably be the easiest, as well as adding the improper potential to the torsion potential.

**Comment by jchodera** [comment]
Are you using OpenMM energies for the torsion drive, or computing the torsions yourself?

**Comment by pgrinaway** [comment]
Computing them myself. I have some code to do this with OpenMM though (needs to be finished), so I could bring that in and finish it in the same PR.

**Comment by jchodera** [comment]
OK, cool.  Perhaps the first thing to fix would be #110 (remove `receptor_topology` argument, move small molecule `resname` to constructor) in a quick PR, then tackle the generalization to topology-based torsions in a separate PR?

**Comment by pgrinaway** [comment]
Yeah, agreed.

**Comment by pgrinaway** [comment]
fixed in #127
